### PR TITLE
[WIP] Measure multithreaded solveequihash time per-thread

### LINKED ISSUE #1191: Make the solveequihash benchmark more representative of solution latency / add another benchmark for throughput
@Vorksholk wrote at https://github.com/zcash/zcash/pull/1174#issuecomment-238038073 :

> Since this benchmark runs solvers on x threads but waits until all threads have finished before starting the next batch of solvers, it doesn't seem to be terribly representative of real-world throughput, since towards the end of the benchmark computational resources are sitting idle.
> 
> For example, running the zcbenchmark on 6 threads for two cycles performs the first cycle, and doesn't start the 2nd cycle until all threads of the first cycle have finished. On my 6-core system, I end up waiting ~10-15 seconds for the final thread to finish its work.
> 
> A 'simple' fix would be modifying the benchmark to solve a "pool" of x equihashes and allow threads to draw from the pool of equihashes on-demand, so that longer benchmarks would come closer and closer to approximating real-world performance since the only idle resources would exist when the pool is empty and some threads have finished their work.

### PR DESCRIPTION
**Pull request by str4d**
Closes #1191

### CODE REVIEW

**Comment by str4d** [comment]
Compilation currently fails with:

```
libbitcoin_wallet.a(libbitcoin_wallet_a-zcbenchmarks.o): In function `boost::system::make_error_code(boost::future_errc)':
/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/thread/future_error_code.hpp:49: undefined reference to `boost::future_category()'
/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/thread/future_error_code.hpp:49: undefined reference to `boost::future_category()'
/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/thread/future_error_code.hpp:49: undefined reference to `boost::future_category()'
/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/thread/future_error_code.hpp:49: undefined reference to `boost::future_category()'
/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/thread/future_error_code.hpp:49: undefined reference to `boost::future_category()'
collect2: error: ld returned 1 exit status
Makefile:2995: recipe for target 'zcash-gtest' failed
```

File: src/zcbenchmarks.cpp [hunk 1]
```diff
@@ -117,35 +119,39 @@ double benchmark_solve_equihash(bool time)
...
+    std::vector<double> ret;
+    std::vector<boost::shared_future<double>> tasks;
+    for (int i = 0; i < nThreads; i++) {
+        tasks.emplace_back(boost::async(&benchmark_solve_equihash));
+    }
+    auto it = boost::wait_for_any(tasks.begin(),tasks.end());
+    while(it != tasks.end()) {
+        ret.push_back(it->get());
+        tasks.erase(it);
+        it = boost::wait_for_any(tasks.begin(),tasks.end());
```
Comment (daira) [inline src/zcbenchmarks.cpp]: This is a slightly strange API (in what it uses iterators for, essentially as just a kind of pointer), but I looked at [its documentation](http://www.boost.org/doc/libs/1_46_0/doc/html/thread/synchronization.html#thread.synchronization.futures.reference.wait_for_any) and I guess this usage is correct.

**Comment by daira** [comment]
Make sure we're linking with ``-lboost_system`` and ``-lboost_thread``.

http://stackoverflow.com/questions/9723793/undefined-reference-to-boostsystemsystem-category-when-compiling

**Comment by str4d** [comment]
It definitely looks like we are linking to both of those. Here's the full make output (after a previous run to compile everything that does compile):

```make
$ make -j3 V=1
Making all in src
make[1]: Entering directory '/home/str4d/dev/zcash/zcash/src'
make[2]: Entering directory '/home/str4d/dev/zcash/zcash/src'
make  -C secp256k1 libsecp256k1.la
make[3]: Entering directory '/home/str4d/dev/zcash/zcash/src/secp256k1'
make[3]: Leaving directory '/home/str4d/dev/zcash/zcash/src/secp256k1'
/bin/bash ../libtool  --tag=CXX   --mode=link /home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/native/bin/ccache g++ -m64  -std=c++11 -pipe -O2 -Wno-deprecated-declarations -Wno-placement-new -Wno-terminate -Werror -O1 -g  -Wstack-protector -fstack-protector-all -fPIE  -pthread   -static -L/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/lib   -Wl,-z,relro -Wl,-z,now -pie -o test/test_bitcoin test/test_test_bitcoin-arith_uint256_tests.o test/test_test_bitcoin-addrman_tests.o test/test_test_bitcoin-alert_tests.o test/test_test_bitcoin-allocator_tests.o test/test_test_bitcoin-base32_tests.o test/test_test_bitcoin-base58_tests.o test/test_test_bitcoin-base64_tests.o test/test_test_bitcoin-bip32_tests.o test/test_test_bitcoin-bloom_tests.o test/test_test_bitcoin-checkblock_tests.o test/test_test_bitcoin-Checkpoints_tests.o test/test_test_bitcoin-coins_tests.o test/test_test_bitcoin-compress_tests.o test/test_test_bitcoin-crypto_tests.o test/test_test_bitcoin-DoS_tests.o test/test_test_bitcoin-equihash_tests.o test/test_test_bitcoin-getarg_tests.o test/test_test_bitcoin-hash_tests.o test/test_test_bitcoin-key_tests.o test/test_test_bitcoin-main_tests.o test/test_test_bitcoin-mempool_tests.o test/test_test_bitcoin-miner_tests.o test/test_test_bitcoin-mruset_tests.o test/test_test_bitcoin-multisig_tests.o test/test_test_bitcoin-netbase_tests.o test/test_test_bitcoin-pmt_tests.o test/test_test_bitcoin-policyestimator_tests.o test/test_test_bitcoin-pow_tests.o test/test_test_bitcoin-reverselock_tests.o test/test_test_bitcoin-rpc_tests.o test/test_test_bitcoin-sanity_tests.o test/test_test_bitcoin-scheduler_tests.o test/test_test_bitcoin-script_P2SH_tests.o test/test_test_bitcoin-script_tests.o test/test_test_bitcoin-scriptnum_tests.o test/test_test_bitcoin-serialize_tests.o test/test_test_bitcoin-sighash_tests.o test/test_test_bitcoin-sigopcount_tests.o test/test_test_bitcoin-skiplist_tests.o test/test_test_bitcoin-test_bitcoin.o test/test_test_bitcoin-timedata_tests.o test/test_test_bitcoin-transaction_tests.o test/test_test_bitcoin-uint256_tests.o test/test_test_bitcoin-univalue_tests.o test/test_test_bitcoin-util_tests.o test/test_test_bitcoin-sha256compress_tests.o test/test_test_bitcoin-accounting_tests.o wallet/test/test_test_bitcoin-wallet_tests.o test/test_test_bitcoin-rpc_wallet_tests.o     libbitcoin_server.a libbitcoin_cli.a libbitcoin_common.a libbitcoin_util.a crypto/libbitcoin_crypto.a univalue/libbitcoin_univalue.a ./leveldb/libleveldb.a ./leveldb/libmemenv.a -L/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/lib -lboost_system-mt -lboost_filesystem-mt -lboost_program_options-mt -lboost_thread-mt -lboost_chrono-mt -lboost_unit_test_framework-mt secp256k1/libsecp256k1.la libbitcoin_wallet.a libbitcoinconsensus.la -ldb_cxx-4.8 -L/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/lib -lssl -lcrypto -ldl -L/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/lib -lcrypto -ldl -lminiupnpc libzcash.a -lsnark -lgmp -lgmpxx -lboost_system-mt -lcrypto -lsodium -fopenmp -lcrypto -lanl 
/bin/bash ../libtool  --tag=CXX   --mode=link /home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/native/bin/ccache g++ -m64  -std=c++11 -pipe -O2 -Wno-deprecated-declarations -Wno-placement-new -Wno-terminate -Werror -O1 -g  -Wstack-protector -fstack-protector-all -fPIE  -pthread   -static -L/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/lib   -Wl,-z,relro -Wl,-z,now -pie -o zcash-gtest gtest/zcash_gtest-main.o gtest/zcash_gtest-test_tautology.o gtest/zcash_gtest-test_checktransaction.o gtest/zcash_gtest-test_equihash.o gtest/zcash_gtest-test_joinsplit.o gtest/zcash_gtest-test_noteencryption.o gtest/zcash_gtest-test_merkletree.o gtest/zcash_gtest-test_circuit.o gtest/zcash_gtest-test_txid.o gtest/zcash_gtest-test_libzcash_utils.o -lgtest -lgmock libbitcoin_server.a libbitcoin_cli.a libbitcoin_common.a libbitcoin_util.a crypto/libbitcoin_crypto.a univalue/libbitcoin_univalue.a ./leveldb/libleveldb.a ./leveldb/libmemenv.a -L/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/lib -lboost_system-mt -lboost_filesystem-mt -lboost_program_options-mt -lboost_thread-mt -lboost_chrono-mt -lboost_unit_test_framework-mt secp256k1/libsecp256k1.la libbitcoin_wallet.a libbitcoinconsensus.la -ldb_cxx-4.8 -L/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/lib -lssl -lcrypto -ldl -L/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/lib -lcrypto -ldl -lminiupnpc libzcash.a -lsnark -lgmp -lgmpxx -lboost_system-mt -lcrypto -lsodium -fopenmp -lcrypto -lanl 
libtool: link: /home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/native/bin/ccache g++ -m64 -std=c++11 -pipe -O2 -Wno-deprecated-declarations -Wno-placement-new -Wno-terminate -Werror -O1 -g -Wstack-protector -fstack-protector-all -fPIE -pthread -Wl,-z -Wl,relro -Wl,-z -Wl,now -pie -o zcash-gtest gtest/zcash_gtest-main.o gtest/zcash_gtest-test_tautology.o gtest/zcash_gtest-test_checktransaction.o gtest/zcash_gtest-test_equihash.o gtest/zcash_gtest-test_joinsplit.o gtest/zcash_gtest-test_noteencryption.o gtest/zcash_gtest-test_merkletree.o gtest/zcash_gtest-test_circuit.o gtest/zcash_gtest-test_txid.o gtest/zcash_gtest-test_libzcash_utils.o -fopenmp  -L/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/lib -lgtest -lgmock libbitcoin_server.a libbitcoin_cli.a libbitcoin_common.a libbitcoin_util.a crypto/libbitcoin_crypto.a univalue/libbitcoin_univalue.a ./leveldb/libleveldb.a ./leveldb/libmemenv.a -lboost_filesystem-mt -lboost_program_options-mt -lboost_thread-mt -lboost_chrono-mt -lboost_unit_test_framework-mt secp256k1/.libs/libsecp256k1.a libbitcoin_wallet.a ./.libs/libbitcoinconsensus.a -ldb_cxx-4.8 -lssl -ldl -lminiupnpc libzcash.a -lsnark /home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/lib/libgmpxx.a /home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/lib/libgmp.a -lboost_system-mt /home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/lib/libsodium.a -lcrypto -lanl -fopenmp -pthread
libtool: link: /home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/native/bin/ccache g++ -m64 -std=c++11 -pipe -O2 -Wno-deprecated-declarations -Wno-placement-new -Wno-terminate -Werror -O1 -g -Wstack-protector -fstack-protector-all -fPIE -pthread -Wl,-z -Wl,relro -Wl,-z -Wl,now -pie -o test/test_bitcoin test/test_test_bitcoin-arith_uint256_tests.o test/test_test_bitcoin-addrman_tests.o test/test_test_bitcoin-alert_tests.o test/test_test_bitcoin-allocator_tests.o test/test_test_bitcoin-base32_tests.o test/test_test_bitcoin-base58_tests.o test/test_test_bitcoin-base64_tests.o test/test_test_bitcoin-bip32_tests.o test/test_test_bitcoin-bloom_tests.o test/test_test_bitcoin-checkblock_tests.o test/test_test_bitcoin-Checkpoints_tests.o test/test_test_bitcoin-coins_tests.o test/test_test_bitcoin-compress_tests.o test/test_test_bitcoin-crypto_tests.o test/test_test_bitcoin-DoS_tests.o test/test_test_bitcoin-equihash_tests.o test/test_test_bitcoin-getarg_tests.o test/test_test_bitcoin-hash_tests.o test/test_test_bitcoin-key_tests.o test/test_test_bitcoin-main_tests.o test/test_test_bitcoin-mempool_tests.o test/test_test_bitcoin-miner_tests.o test/test_test_bitcoin-mruset_tests.o test/test_test_bitcoin-multisig_tests.o test/test_test_bitcoin-netbase_tests.o test/test_test_bitcoin-pmt_tests.o test/test_test_bitcoin-policyestimator_tests.o test/test_test_bitcoin-pow_tests.o test/test_test_bitcoin-reverselock_tests.o test/test_test_bitcoin-rpc_tests.o test/test_test_bitcoin-sanity_tests.o test/test_test_bitcoin-scheduler_tests.o test/test_test_bitcoin-script_P2SH_tests.o test/test_test_bitcoin-script_tests.o test/test_test_bitcoin-scriptnum_tests.o test/test_test_bitcoin-serialize_tests.o test/test_test_bitcoin-sighash_tests.o test/test_test_bitcoin-sigopcount_tests.o test/test_test_bitcoin-skiplist_tests.o test/test_test_bitcoin-test_bitcoin.o test/test_test_bitcoin-timedata_tests.o test/test_test_bitcoin-transaction_tests.o test/test_test_bitcoin-uint256_tests.o test/test_test_bitcoin-univalue_tests.o test/test_test_bitcoin-util_tests.o test/test_test_bitcoin-sha256compress_tests.o test/test_test_bitcoin-accounting_tests.o wallet/test/test_test_bitcoin-wallet_tests.o test/test_test_bitcoin-rpc_wallet_tests.o -fopenmp  -L/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/lib libbitcoin_server.a libbitcoin_cli.a libbitcoin_common.a libbitcoin_util.a crypto/libbitcoin_crypto.a univalue/libbitcoin_univalue.a ./leveldb/libleveldb.a ./leveldb/libmemenv.a -lboost_filesystem-mt -lboost_program_options-mt -lboost_thread-mt -lboost_chrono-mt -lboost_unit_test_framework-mt secp256k1/.libs/libsecp256k1.a libbitcoin_wallet.a ./.libs/libbitcoinconsensus.a -ldb_cxx-4.8 -lssl -ldl -lminiupnpc libzcash.a -lsnark /home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/lib/libgmpxx.a /home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/lib/libgmp.a -lboost_system-mt /home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/lib/libsodium.a -lcrypto -lanl -fopenmp -pthread
libbitcoin_wallet.a(libbitcoin_wallet_a-zcbenchmarks.o): In function `boost::system::make_error_code(boost::future_errc)':
/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/thread/future_error_code.hpp:49: undefined reference to `boost::future_category()'
/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/thread/future_error_code.hpp:49: undefined reference to `boost::future_category()'
/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/thread/future_error_code.hpp:49: undefined reference to `boost::future_category()'
/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/thread/future_error_code.hpp:49: undefined reference to `boost::future_category()'
/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/thread/future_error_code.hpp:49: undefined reference to `boost::future_category()'
collect2: error: ld returned 1 exit status
Makefile:2995: recipe for target 'zcash-gtest' failed
make[2]: *** [zcash-gtest] Error 1
make[2]: *** Waiting for unfinished jobs....
libbitcoin_wallet.a(libbitcoin_wallet_a-zcbenchmarks.o): In function `boost::system::make_error_code(boost::future_errc)':
/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/thread/future_error_code.hpp:49: undefined reference to `boost::future_category()'
/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/thread/future_error_code.hpp:49: undefined reference to `boost::future_category()'
/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/thread/future_error_code.hpp:49: undefined reference to `boost::future_category()'
/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/thread/future_error_code.hpp:49: undefined reference to `boost::future_category()'
/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/thread/future_error_code.hpp:49: undefined reference to `boost::future_category()'
collect2: error: ld returned 1 exit status
Makefile:2961: recipe for target 'test/test_bitcoin' failed
make[2]: *** [test/test_bitcoin] Error 1
make[2]: Leaving directory '/home/str4d/dev/zcash/zcash/src'
Makefile:7075: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/home/str4d/dev/zcash/zcash/src'
Makefile:644: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1
```

**Comment by defuse** [comment]
ACK (not sure why the build fails, the order of the `-l` flags does matter so could it be that?).

**Comment by daira** [comment]
Booting this out of z9.

**Comment by ebfull** [comment]
Can this be rebased or closed? @str4d

**Comment by str4d** [comment]
I'll rebase this, and if we can figure out what the compile issue is, it can be merged.

**Comment by ebfull** [comment]
Any update? :)

**Comment by str4d** [comment]
Still not working, and I still can't figure out why.

**Comment by daira** [comment]
Is the compilation problem specific to your machine @str4d?

**Comment by daira** [comment]
I guess @zkbot try will tell me that.

**Comment by ebfull** [comment]
I've closed this because we've switched to the `master` branch and reset homu.

**Pull request closed without merging**